src/Altinn.Platform/Altinn.Platform.Authorization/Authorization/Services/Implementation/PolicyInformationPoint.cs

```diff
@@ -0,0 +1,127 @@
+using Altinn.Authorization.ABAC.Xacml;
+using Altinn.Platform.Authorization.Configuration;
+using Altinn.Platform.Authorization.Helpers;
+using Altinn.Platform.Authorization.Models;
+using Altinn.Platform.Authorization.Repositories;
+using Altinn.Platform.Authorization.Repositories.Interface;
+using Altinn.Platform.Authorization.Services.Interface;
+using Azure;
+using Azure.Storage.Blobs.Models;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Extensions.Options;
+
+namespace Altinn.Platform.Authorization.Services.Implementation
+{
+    /// <summary>
+    /// The Policy Administration Point responsible for storing and modifying delegation policies
+    /// </summary>
+    public class PolicyInformationPoint : IPolicyInformationPoint
+    {
+        private readonly IPolicyRetrievalPoint _prp;
+        private readonly IDelegationMetadataRepository _delegationRepository;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PolicyAdministrationPoint"/> class.
+        /// </summary>
+        /// <param name="policyRetrievalPoint">The policy retrieval point</param>
+        /// <param name="policyRepository">The policy repository</param>
+        /// <param name="delegationRepository">The delegation change repository</param>
+        /// <param name="memoryCache">The cache handler</param>
+        /// <param name="settings">The app settings</param>
+        public PolicyInformationPoint(IPolicyRetrievalPoint policyRetrievalPoint, IPolicyRepository policyRepository, IDelegationMetadataRepository delegationRepository, IMemoryCache memoryCache, IOptions<GeneralSettings> settings)
+        {
+            _prp = policyRetrievalPoint;
+            _delegationRepository = delegationRepository;
+        }
+
+        /// <inheritdoc/>
+        public async Task<List<Rule>> GetRulesAsync(List<string> orgApp, List<int> offeredByPartyIds, List<int> coveredByPartyIds, List<int> coveredByUserIds)
+        {
+            List<Rule> rules = new List<Rule>();
+            List<DelegationChange> delegationChanges = await _delegationRepository.GetAllCurrentDelegationChanges(orgApp, offeredByPartyIds, coveredByPartyIds, coveredByUserIds);
+            foreach (DelegationChange delegationChange in delegationChanges)
+            {
+                XacmlPolicy policy = await _prp.GetPolicyVersionAsync(delegationChange.BlobStoragePolicyPath, delegationChange.BlobStorageVersionId);
+                rules.AddRange(GetRulesFromPolicyAndDelegationChange(policy.Rules, delegationChange));
+            }
+
+            return rules;
+        }
+
+        private static List<Rule> GetRulesFromPolicyAndDelegationChange(ICollection<XacmlRule> xacmlRules, DelegationChange delegationChange)
+        {
+            List<Rule> rules = new List<Rule>();
+            foreach (XacmlRule xacmlRule in xacmlRules)
+            {
+                if (xacmlRule.Effect.Equals(XacmlEffectType.Permit) && xacmlRule.Target != null)
```

Comment:
Here we check that `Target` has a value, but it isn't used by the method. Upon further inspection we see it's used by the `AddAttributeMatchesToRule` method. I suggest changing the parameter type to `AddAttributeMatchesToRule` so that it's input is `Target`. That way the check is more logically placed.

Reply:
Changed AddAttributeMatchesToRule's parameter type from XacmlRule to XacmlTarget

Reply:
Changed AddAttributeMatchesToRule's parameter type from XacmlRule to XacmlTarget